src/TorchSharp/DisposeScopeManager.cs

```diff
@@ -0,0 +1,315 @@
+// Copyright (c) .NET Foundation and Contributors.  All Rights Reserved.  See LICENSE in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Threading;
+
+namespace TorchSharp
+{
+    /// <summary>
+    /// Manages dispose scopes, that can make automatic tensor disposal easier. Note that the
+    /// DisposeManager is thread local. The DisposeScopeManager can also manage other disposables, such as training
+    /// batches and the like.
+    /// </summary>
+    public class DisposeScopeManager
+    {
+        private int _createdCount;
+        private int _disposedCount;
+
+        static DisposeScopeManager()
+        {
+            torch.Tensor.OnTensorCreateCallback.Add(TryRegisterInDisposeScope);
+            torch.Tensor.OnTensorDisposeCallback.Add(TryUnregisterInDisposeScope);
+        }
+
+        [ThreadStatic] private static DisposeScopeManager _singleton;
+
+        public static DisposeScopeManager Singleton = (_singleton ??= new DisposeScopeManager());
+
+        internal Stack<DisposeScope> DisposeScopeStack { get; } = new();
+        internal IDisposable CurrentlyDisposing { get; set; }
+
+        public int CreatedCount => _createdCount;
+        public int DisposedCount => _disposedCount;
+        public int TotalCount => _createdCount - _disposedCount;
```

Comment:
NIT: LiveCount ?

Reply:
It's because of this naming on Tensor;

     public static long TotalCount => _totalCount;

I figured I'd use the same name?

Reply:
I could call it a Threadaton? But you're probably right, I'll call  it ThreadSingleton so that it becomes apparant that it's not really a singleton.

Reply:
Maybe `ForCurrentThread`

Reply:
Fixed with

    /// <summary>
    /// Keeps track of statistics for a dispose scope. Can be queried to figure out performance/memory issues.
    /// </summary>
    public class ThreadDisposeScopeStatistics
    {
        /// <summary>
        /// The number of disposables that were created on this thread, but weren't captured by a DisposeScope.
        /// </summary>
        public long CreatedOutsideScopeCount { get; internal set; }

        /// <summary>
        /// The number of disposables that were created on this thread and were captured by a DisposeScope.
        /// </summary>
        public long CreatedInScopeCount { get; internal set; }

        /// <summary>
        /// The number of disposables that were disposed on this thread and were disposed while in a DisposeScope.
        /// </summary>
        public long DisposedInScopeCount { get; internal set; }

        /// <summary>
        /// Number of disposables that were once included in the scope, but were subsequently detached.
        /// </summary>
        public long DetachedFromScopeCount { get; internal set; }

        /// <summary>
        /// The number of disposables that are currently live on the current thread. If a It's aproximate, see
        /// Tensor.TotalCount. Disposables that weren't created within a DisposeScope, or detached from the dispose
        /// scope, will not be counted as alive.
        /// </summary>
        public long ThreadTotalLiveCount => CreatedInScopeCount - DisposedInScopeCount - DetachedFromScopeCount;

        /// <summary>
        /// Resets the counts for the current thread. See ThreadTotalLiveCount etc. Mainly used in tests to make sure
        /// we get a clean slate on the thread.
        /// </summary>
        public void Reset()
        {
            CreatedOutsideScopeCount = 0;
            CreatedInScopeCount = 0;
            DisposedInScopeCount = 0;
            DetachedFromScopeCount = 0;
        }
    }